### PR TITLE
Frontend for Uploading a Snapshot 

### LINKED ISSUE #3866: Add front end API and UI for Upload snapshot

### PR DESCRIPTION
**Pull request by Steve-Mcl**
Task 2 of 3 (part of #3628)

## Description

Adds UI parts for Uploading a Snapshot 

### Details
- The "Upload Snapshot" buttons are added to the instance and device pages (not the application page: reason provided on the issue)
  - The buttons placed adjacent to the "Create Snapshot" buttons on both the table search row AND the empty state actions slot
  - The buttons are only visible to users with the `snapshots:import` role
  - Fixes the `gap` between the buttons in the empty state actions slot (in line with the SCSS `gap` setting for the table search actions slot)
- Adds shared component `frontend/src/components/dialogs/SnapshotImportDialog.vue`
  - this is reused on both the instance and device pages
- Adds a utility function `isSnapshot` in `forge/routes/api/snapshot.js`
  - this is used to do basic validation the snapshot before attempting to import it (browser does not have native support for JSON schema validation, so this is a simple check to ensure the snapshot has the required properties and are of the correct type)
- Adds audit log entry for the import action
  - corrects icon for snapshot log entries
- Adds E2E tests for the new endpoint for both instances and devices

 
### NOTE 
- Backend and Docs are in separate PRs (see parent story #3628)

### E2E tests added
```
  FlowForge - Devices - With Billing
    upload snapshot with credentials
    upload snapshot without credentials
  FlowForge - Instance Snapshots
    upload snapshot with credentials
    upload snapshot without credentials
  FlowForge shows audit logs
    for when a snapshot is imported
```

## Related Issue(s)

Closes #3866
Story: #3628 

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by joepavitt** [comment]
I think the order of the buttons should be this way arund:

Screenshot 2024-05-16 at 14 10 36

The primary option being on the right is a better standard to go by, the "Upload Snapshot" will be less-used

File: frontend/src/components/dialogs/SnapshotImportDialog.vue [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+                <FormRow :error="validateField.file ? errors.file : ''" data-form="import-snapshot-filename">
+                    Snapshot File
+                    <template #input>
+                        <ff-text-input v-model="input.file" :error="errors.file" disabled />
+                        <!-- <div class="ff-input ff-text-input">
+                            <input type="text" readonly class="bg-gray-50 border border-gray-300  focus:outline-none text-gray-500 cursor-not-allowed" :value="input.file">
+                        </div> -->
+                    </template>
+                    <template #append>
+                        <input id="fileUpload" ref="fileUpload" type="file" accept="application/json, text/plain, *" class="" style="/*position: absolute; top: -9999px;*/ display:none;">
```
Comment (joepavitt) [inline frontend/src/components/dialogs/SnapshotImportDialog.vue:14]: commented out inline `style`?

Comment (Steve-Mcl) [reply frontend/src/components/dialogs/SnapshotImportDialog.vue:14]: ah yes, before I figured out how to cypress test a hidden file input.

This is what it should be ↓

```suggestion
                        <input id="fileUpload" ref="fileUpload" type="file" accept="application/json, text/plain, *" style="display:none;">
```

File: frontend/src/components/dialogs/SnapshotImportDialog.vue [hunk 2]
```diff
@@ -0,0 +1,208 @@
+<template>
+    <ff-dialog ref="dialog" :header="title" confirm-label="Upload" :disable-primary="!formValid" :closeOnConfirm="false" @confirm="confirm()" @cancel="cancel">
+        <template #default>
+            <form class="space-y-6 mt-2" @submit.prevent>
+                <FormRow :error="validateField.file ? errors.file : ''" data-form="import-snapshot-filename">
+                    Snapshot File
+                    <template #input>
+                        <ff-text-input v-model="input.file" :error="errors.file" disabled />
+                        <!-- <div class="ff-input ff-text-input">
+                            <input type="text" readonly class="bg-gray-50 border border-gray-300  focus:outline-none text-gray-500 cursor-not-allowed" :value="input.file">
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/components/dialogs/SnapshotImportDialog.vue:10]: yup, last minute stuff from getting cypress tests working at about 9:45 PM last night 🤦‍♂️

File: frontend/src/components/dialogs/SnapshotImportDialog.vue [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                        </div> -->
```
Comment (Steve-Mcl) [inline frontend/src/components/dialogs/SnapshotImportDialog.vue:11]:
```suggestion
```

File: frontend/src/components/dialogs/SnapshotImportDialog.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/components/dialogs/SnapshotImportDialog.vue]: removed now.

**Comment by Steve-Mcl** [comment]
> I think the order of the buttons should be this way arund:
> 
> Screenshot 2024-05-16 at 14 10 36


Resolved in [swap upload and create buttons on both empty-state and search toolbar](https://github.com/FlowFuse/flowfuse/pull/3869/commits/2e296a6cd04f9b92706240e4d5de230d6ff4bf3d) [2e296a6](https://github.com/FlowFuse/flowfuse/pull/3869/commits/2e296a6cd04f9b92706240e4d5de230d6ff4bf3d)

**Comment by Steve-Mcl** [comment]
@joepavitt backend part is now merged (I have updated this PR too) so you can merge upon approval. Ta.

**Review by joepavitt — APPROVED** [review]
Approved subject to final tests passing, which they have done previously

**Pull request merged**